src/main/java/formflow/library/ScreenController.java

```diff
@@ -131,6 +131,7 @@ ModelAndView postScreen(
     // if there's already a session
     if (submission.getId() != null) {
       submission.mergeFormDataWithSubmissionData(formSubmission);
+      handleBeforeSaveAction(getScreenConfig(flow, screen), submission);
       saveToRepository(submission);
```

Comment:
Thinking bigger picture here. We allow the user to do whatever they want with the data before they save it.  This is _after_ we validate the data. So we know the initial data is good, but you could introduce all sorts of invalid data via this hook.   A little bit of a yellow flag for me.  But presumably the app writer knows what they are doing. 

Also, in the case of the date action, it allows you to set the date, but not actually validate that it's a good / real one, unless the action hook does it itself.  The fields have all been validated so far, so hopefully that's good and solid, but you have no sense if the aggregated Date field is solid. 

It's making me think that we need a pre-validation hook to manipulate the data. Then we could have (in the FlowInputs Class (UBI.java)) separate fields -- day, month, year, Date. Then Date could be validated as well.  This could solve some issues we are facing with the custom validations. 

If we have this pre-validation action hook, we could check multiple fields and set flags in the object. Then during validation, if the flag doesn't check out, we trigger an error.

Reply:
Yeah that makes sense